domain/cloudimagemetadata/service/service.go

```diff
@@ -0,0 +1,129 @@
+func NewService(st State) *Service {
+	return &Service{
+		st: st,
+	}
+}
+
+// SaveMetadata saves the provided cloud image metadata if non-empty and valid.
+// It returns a [errors.NothingToSave] if the input are empty, or a [errors.Invalid] if at least one of the inputs are
+// invalid.
+func (s Service) SaveMetadata(ctx context.Context, metadata []cloudimagemetadata.Metadata) error {
+	if len(metadata) == 0 {
+		return errors.NothingToSave
+	}
+
+	if err := s.validateAllMetadata(metadata); err != nil {
+		return err
+	}
+
+	return s.st.SaveMetadata(ctx, metadata)
+}
+
+// DeleteMetadataWithImageID removes all the metadata associated with the given imageID from the state.
+func (s Service) DeleteMetadataWithImageID(ctx context.Context, imageId string) error {
+	return s.st.DeleteMetadataWithImageID(ctx, imageId)
+}
+
+// FindMetadata retrieves a map of image metadata grouped by the source based on the provided filter criteria.
+func (s Service) FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) (map[string][]cloudimagemetadata.Metadata, error) {
+	metadata, err := s.st.FindMetadata(ctx, criteria)
+	if err != nil {
+		return nil, err
+	}
+	metadataBySource := make(map[string][]cloudimagemetadata.Metadata)
+	for _, v := range metadata {
+		metadataBySource[v.Source] = append(metadataBySource[v.Source], v)
+	}
+	return metadataBySource, nil
+}
+
+// AllCloudImageMetadata retrieves all cloud image metadata from the state and returns them as a list.
+func (s Service) AllCloudImageMetadata(ctx context.Context) ([]cloudimagemetadata.Metadata, error) {
+	return s.st.AllCloudImageMetadata(ctx)
+}
+
+// validateAllMetadata validates a list of cloud image metadata.
+// It returns an error if any of the metadata entries are invalid.
+func (s Service) validateAllMetadata(metadata []cloudimagemetadata.Metadata) error {
+	var errs []error
+	for _, m := range metadata {
+		errs = append(errs, s.validateMetadata(m))
+	}
+	return jujuerrors.Join(errs...)
+}
+
+// validateMetadata validates a single cloud image metadata entry.
+func (s Service) validateMetadata(m cloudimagemetadata.Metadata) error {
+	if m.ImageID == "" {
+		return jujuerrors.Errorf("image id is empty: %w", errors.Invalid)
+	}
+
+	var missing []string
+	if m.Version == "" {
+		missing = append(missing, "version")
+	}
+	if m.Stream == "" {
+		missing = append(missing, "stream")
+	}
+	if m.Source == "" {
+		missing = append(missing, "source")
+	}
+	if m.Arch == "" {
+		missing = append(missing, "arch")
+	}
+	if m.Region == "" {
+		missing = append(missing, "region")
+	}
+
+	if len(missing) > 0 {
+		return errors.NotValid(m.ImageID, missing)
+	}
+
+	supportedArchitectures := s.st.SupportedArchitectures(context.Background())
+	if !supportedArchitectures.Contains(m.Arch) {
+		return jujuerrors.Errorf("unsupported architecture %s (should be any of %s): %w", m.Arch, supportedArchitectures.Values(), errors.Invalid)
```

Comment:
Just to be sure, `errors.NotValid(..)` returns an error that satisfies `errors.Is(err, errors.Invalid)` right? If not, why not using `errors.NotValidf(...`?

Reply:
It returns an error who satisfies `github.com/juju/juju/domain/cloudimagemetadata/errors.Invalid`.

I don't know from which errors package you get `error.NotValidf` but I assumes it comes from `github.com/juju/juju/errors` which we don't use anymore, as far i understand, in favor of domain-scoped errors.

Reply:
Oh I see, it's from your domain errors' package. That's fine then. IMO you should name the error `NotValid` as well, just for consistency because that's what we have all around our codebase.